docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+**Pros**:
+
+* This is the least-intrusive approach.
+
+* It will work off the box without requiring any additional configuration.
+
+**Cons**:
+
+* `du` can consume a lot of cpu and memory. There have been several issues reported against the kubelet in the past that were related to `du`.
+
+* It is time consuming. Cannot be run frequently. Requires special handling to constrain resource usage - setting lower nice value or running in a sub-container.
+
+* Can block container deletion by keeping file descriptors open.
+
+
+2. Linux gid based Disk Quota
+
+	[Disk quota](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Storage_Administration_Guide/ch-disk-quotas.html) feature provided by the linux kernel can be used to track the usage of image layers. Ideally, we need `project` support for disk quota, which let’s us track usage of directory hierarchies using `project ids`. Unfortunately, that feature is only available for zfs filesystems. Since most of our distributions use `ext4` by default, we will have to use either `uid` or `gid` based quota tracking.
+
+	Both `uids` and `gids` are meant for security. Overloading that concept for disk tracking is painful and ugly. But, that is what we have today.
+
+	Kubelet needs to define a gid for tracking image layers and make that gid or group the owner of `/var/lib/docker/[aufs | overlayfs]` recursively. Once this is done, the quota sub-system in the kernel will report the blocks being consumed by the storage driver on the underlying partition.
+
+	Since this number also includes the container’s writable layer, we will have to somehow subtract that usage from the overall usage of the storage driver directory. Luckily, we can use the same mechanism for tracking container’s writable layer. Once we apply a different `gid` to the container’s writable layer, which is located under `/var/lib/docker/<storage_driver>/diff/<container_id>`, the quota subsystem will not include the container’s writable layer usage.
+
+**Pros**:
+
+* Low overhead tracking provided by the kernel.
+
+
+**Cons**
+
+* Requires updates to default ownership on docker’s internal storage driver directories. We will have to deal with storage driver implementation implementation details in any approach that is not docker native.
+
+* Requires additional node configuration - quota subsystem needs to be setup on the node. This can either be automated or made a requirement for the node.
+
+* Kubelet needs to perform gid management. A range of gids have to allocated to the kubelet for the purposes of quota management. This range must not be used for any other purposes out of band.
+
+* Breaks `docker save` semantics. Since kubernetes assumes immutable images, this is not a blocker. To support quota in docker, we will need user-namespaces along with custom gid mapping for each container. This feature does not exist today.
```

Comment:
> Depending on how we view the problem, not sharing images might actually be a feature. We would ideally want to safeguard the node from a container that pulls a large image and ends up affecting all running containers. How about using a volume to share images across these `dind` containers?

> @ncdc: How does CI work today with Open Shift? Does it still use the host docker daemon?

We actually stopped using dind because it was too problematic at the time. If we were to go back to it, the volume sharing idea is interesting. I'm not sure how practical it would be, since the Docker daemon requires exclusive read/write access to its storage, but maybe there'd be some way to cache and reuse the directory for each new dind pod. Or further down the road, having the image graph storage use shared storage (ceph/gluster/etc) would be useful.

As for how OpenShift does builds, we bind mount the node's Docker socket into the build pod's container, so that it can talk to Docker to create a new container to perform the build. This is problematic as it bypasses the scheduler and can potentially have negative impacts to the health and stability of a node.